### PR TITLE
Add empty params check for unused prop types rule to fix empty proptype functions from causing crashes

### LINKED ISSUE #1542: no-unused-prop-types: Potential regression introduced in 7.5.0
Hi there.

Noticed that my lint step failed after tagging my library with no code changes:

https://travis-ci.org/LINKIWI/react-elemental/builds/304035438
https://travis-ci.org/LINKIWI/react-elemental/builds/304036629

Interestingly, the failing build pulled in 7.5.0 of `eslint-plugin-react` which was published a couple minutes before the build kicked off.

I haven't had time to do thorough investigation yet, but the 7.5.0 release seems to have introduced a regression in `no-unused-prop-types`:

```
Cannot read property 'properties' of undefined

TypeError: Cannot read property 'properties' of undefined

    at markPropTypesAsUsed (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint-plugin-react/lib/rules/no-unused-prop-types.js:658:40)

    at handleSetStateUpdater (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint-plugin-react/lib/rules/no-unused-prop-types.js:955:7)

    at Object.handleFunctionLikeExpressions (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint-plugin-react/lib/rules/no-unused-prop-types.js:964:7)

    at updatedRuleInstructions.(anonymous function) (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint-plugin-react/lib/util/Components.js:654:75)

    at listeners.(anonymous function).forEach.listener (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint/lib/util/safe-emitter.js:47:58)

    at Array.forEach (native)

    at Object.emit (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint/lib/util/safe-emitter.js:47:38)

    at NodeEventGenerator.applySelector (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint/lib/util/node-event-generator.js:251:26)

    at NodeEventGenerator.applySelectors (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint/lib/util/node-event-generator.js:280:22)

    at NodeEventGenerator.enterNode (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint/lib/util/node-event-generator.js:294:14)

    at CodePathAnalyzer.enterNode (/home/travis/build/LINKIWI/react-elemental/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
```

### PR DESCRIPTION
**Pull request by kevinzwhuang**
Summary
----------
There seems to still be an unresolved regression introduced in 7.5.0 from https://github.com/yannickcr/eslint-plugin-react/pull/1507 
(see https://github.com/yannickcr/eslint-plugin-react/issues/1581#issuecomment-351220583)

When passing an empty function as a PropType - it causes the `no-unused-prop-types` to crash eslint with the following error:
```js
Cannot read property 'properties' of undefined
TypeError: Cannot read property 'properties' of undefined
```

When this part of `no-unused-prop-types` is run:
https://github.com/yannickcr/eslint-plugin-react/blob/f6e4c891a474aaff8f69088b174de8c1317f8fa8/lib/rules/no-unused-prop-types.js#L652-L660

it breaks because `node.params` is an empty array. In my testing, it looks like #1507 introduced this breakage (the test I've included does not break when you revert it), it's unclear and hard to determine why #1507 caused this change.

This PR keeps the tests #1507 introduced passing, while also fixing the regression mentioned.

Changes Proposed
-------------------
- [x] Add empty params check in the `markPropTypesAsUsed` function to fix empty propType functions from causing crashes.
- [x] Add test for validating empty propType functions by `no-unused-prop-types`

To reproduce
--------------
Lint this code with `no-unused-prop-types` and it will crash. We're also seeing this live in CI deployments at https://github.com/react-bootstrap/react-bootstrap in tests relating to propTypes.

```es6
class MyComponent extends React.Component<Props> {
  render() {
    return <div>{ this.props.other }</div>
  }
}
MyComponent.propTypes = { other: () => {} };
```

---
Fixes #1542, #1581
cc: @ljharb

### CODE REVIEW

**Comment by kevinzwhuang** [comment]
This fix is similar to https://github.com/yannickcr/eslint-plugin-react/pull/1545

File: tests/lib/rules/no-unused-prop-types.js [hunk 1]
```diff
@@ -2440,6 +2440,17 @@ ruleTester.run('no-unused-prop-types', rule, {
...
+    {
+      code: `
+        class MyComponent extends React.Component<Props> {
+          render() {
+            return <div>{ this.props.other }</div>
+          }
+        }
+        MyComponent.propTypes = { other: () => {} };
+      `,
+      parser: 'babel-eslint'
```
Comment (ljharb) [inline tests/lib/rules/no-unused-prop-types.js]: this shouldn't need babel-eslint; can you either add a duplicate test case (one using the default parser, and one using babel-eslint), or, remove the parser line and only test using the default parser?

Comment (kevinzwhuang) [reply tests/lib/rules/no-unused-prop-types.js]: You're right, I'll take out the parser line and leave the test cases using the default parser

**Comment by kevinzwhuang** [comment]
@ljharb Added test cases for `{ other() {} }`, `{ other: function () { } }`, and `{ * other() {} }`.

Also can confirm the new test cases still crash without the fix:
```es6

  3067 passing (11s)
  4 failing

  1) no-unused-prop-types
       valid

        class MyComponent extends React.Component {
          render() {
            return <div>{ this.props.other }</div>
          }
        }
        MyComponent.propTypes = { other: () => {} };
      :
     TypeError: Cannot read property 'properties' of undefined
      at markPropTypesAsUsed (lib/rules/no-unused-prop-types.js:9:30298)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:39583)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:14919)
      at markPropTypesAsDeclared (lib/rules/no-unused-prop-types.js:9:38650)
      at MemberExpression (lib/rules/no-unused-prop-types.js:9:50264)
      at listeners.(anonymous function).forEach.listener (node_modules/eslint/lib/util/safe-emitter.js:47:58)
      at Array.forEach (native)
      at Object.emit (node_modules/eslint/lib/util/safe-emitter.js:47:38)
      at NodeEventGenerator.applySelector (node_modules/eslint/lib/util/node-event-generator.js:251:26)
      at NodeEventGenerator.applySelectors (node_modules/eslint/lib/util/node-event-generator.js:280:22)
      at NodeEventGenerator.enterNode (node_modules/eslint/lib/util/node-event-generator.js:294:14)
      at CodePathAnalyzer.enterNode (node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
      at Traverser.enter (node_modules/eslint/lib/linter.js:956:32)
      at Traverser.__execute (node_modules/estraverse/estraverse.js:397:31)
      at Traverser.traverse (node_modules/estraverse/estraverse.js:501:28)
      at Traverser.traverse (node_modules/eslint/lib/util/traverser.js:31:22)
      at Linter._verifyWithoutProcessors (node_modules/eslint/lib/linter.js:953:19)
      at preprocess.map.textBlock (node_modules/eslint/lib/linter.js:994:35)
      at Array.map (native)
      at Linter.verify (node_modules/eslint/lib/linter.js:993:42)
      at runRuleForItem (node_modules/eslint/lib/testers/rule-tester.js:366:34)
      at testValidTemplate (node_modules/eslint/lib/testers/rule-tester.js:396:28)
      at Context.RuleTester.it (node_modules/eslint/lib/testers/rule-tester.js:540:25)

  2) no-unused-prop-types
       valid

        class MyComponent extends React.Component {
          render() {
            return <div>{ this.props.other }</div>
          }
        }
        MyComponent.propTypes = { other() {} };
      :
     TypeError: Cannot read property 'properties' of undefined
      at markPropTypesAsUsed (lib/rules/no-unused-prop-types.js:9:30298)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:39583)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:14919)
      at markPropTypesAsDeclared (lib/rules/no-unused-prop-types.js:9:38650)
      at MemberExpression (lib/rules/no-unused-prop-types.js:9:50264)
      at listeners.(anonymous function).forEach.listener (node_modules/eslint/lib/util/safe-emitter.js:47:58)
      at Array.forEach (native)
      at Object.emit (node_modules/eslint/lib/util/safe-emitter.js:47:38)
      at NodeEventGenerator.applySelector (node_modules/eslint/lib/util/node-event-generator.js:251:26)
      at NodeEventGenerator.applySelectors (node_modules/eslint/lib/util/node-event-generator.js:280:22)
      at NodeEventGenerator.enterNode (node_modules/eslint/lib/util/node-event-generator.js:294:14)
      at CodePathAnalyzer.enterNode (node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
      at Traverser.enter (node_modules/eslint/lib/linter.js:956:32)
      at Traverser.__execute (node_modules/estraverse/estraverse.js:397:31)
      at Traverser.traverse (node_modules/estraverse/estraverse.js:501:28)
      at Traverser.traverse (node_modules/eslint/lib/util/traverser.js:31:22)
      at Linter._verifyWithoutProcessors (node_modules/eslint/lib/linter.js:953:19)
      at preprocess.map.textBlock (node_modules/eslint/lib/linter.js:994:35)
      at Array.map (native)
      at Linter.verify (node_modules/eslint/lib/linter.js:993:42)
      at runRuleForItem (node_modules/eslint/lib/testers/rule-tester.js:366:34)
      at testValidTemplate (node_modules/eslint/lib/testers/rule-tester.js:396:28)
      at Context.RuleTester.it (node_modules/eslint/lib/testers/rule-tester.js:540:25)

  3) no-unused-prop-types
       valid

        class MyComponent extends React.Component {
          render() {
            return <div>{ this.props.other }</div>
          }
        }
        MyComponent.propTypes = { other: function () {} };
      :
     TypeError: Cannot read property 'properties' of undefined
      at markPropTypesAsUsed (lib/rules/no-unused-prop-types.js:9:30298)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:39583)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:14919)
      at markPropTypesAsDeclared (lib/rules/no-unused-prop-types.js:9:38650)
      at MemberExpression (lib/rules/no-unused-prop-types.js:9:50264)
      at listeners.(anonymous function).forEach.listener (node_modules/eslint/lib/util/safe-emitter.js:47:58)
      at Array.forEach (native)
      at Object.emit (node_modules/eslint/lib/util/safe-emitter.js:47:38)
      at NodeEventGenerator.applySelector (node_modules/eslint/lib/util/node-event-generator.js:251:26)
      at NodeEventGenerator.applySelectors (node_modules/eslint/lib/util/node-event-generator.js:280:22)
      at NodeEventGenerator.enterNode (node_modules/eslint/lib/util/node-event-generator.js:294:14)
      at CodePathAnalyzer.enterNode (node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
      at Traverser.enter (node_modules/eslint/lib/linter.js:956:32)
      at Traverser.__execute (node_modules/estraverse/estraverse.js:397:31)
      at Traverser.traverse (node_modules/estraverse/estraverse.js:501:28)
      at Traverser.traverse (node_modules/eslint/lib/util/traverser.js:31:22)
      at Linter._verifyWithoutProcessors (node_modules/eslint/lib/linter.js:953:19)
      at preprocess.map.textBlock (node_modules/eslint/lib/linter.js:994:35)
      at Array.map (native)
      at Linter.verify (node_modules/eslint/lib/linter.js:993:42)
      at runRuleForItem (node_modules/eslint/lib/testers/rule-tester.js:366:34)
      at testValidTemplate (node_modules/eslint/lib/testers/rule-tester.js:396:28)
      at Context.RuleTester.it (node_modules/eslint/lib/testers/rule-tester.js:540:25)

  4) no-unused-prop-types
       valid

        class MyComponent extends React.Component {
          render() {
            return <div>{ this.props.other }</div>
          }
        }
        MyComponent.propTypes = { * other() {} };
      :
     TypeError: Cannot read property 'properties' of undefined
      at markPropTypesAsUsed (lib/rules/no-unused-prop-types.js:9:30298)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:39583)
      at iterateProperties (lib/rules/no-unused-prop-types.js:9:14919)
      at markPropTypesAsDeclared (lib/rules/no-unused-prop-types.js:9:38650)
      at MemberExpression (lib/rules/no-unused-prop-types.js:9:50264)
      at listeners.(anonymous function).forEach.listener (node_modules/eslint/lib/util/safe-emitter.js:47:58)
      at Array.forEach (native)
      at Object.emit (node_modules/eslint/lib/util/safe-emitter.js:47:38)
      at NodeEventGenerator.applySelector (node_modules/eslint/lib/util/node-event-generator.js:251:26)
      at NodeEventGenerator.applySelectors (node_modules/eslint/lib/util/node-event-generator.js:280:22)
      at NodeEventGenerator.enterNode (node_modules/eslint/lib/util/node-event-generator.js:294:14)
      at CodePathAnalyzer.enterNode (node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
      at Traverser.enter (node_modules/eslint/lib/linter.js:956:32)
      at Traverser.__execute (node_modules/estraverse/estraverse.js:397:31)
      at Traverser.traverse (node_modules/estraverse/estraverse.js:501:28)
      at Traverser.traverse (node_modules/eslint/lib/util/traverser.js:31:22)
      at Linter._verifyWithoutProcessors (node_modules/eslint/lib/linter.js:953:19)
      at preprocess.map.textBlock (node_modules/eslint/lib/linter.js:994:35)
      at Array.map (native)
      at Linter.verify (node_modules/eslint/lib/linter.js:993:42)
      at runRuleForItem (node_modules/eslint/lib/testers/rule-tester.js:366:34)
      at testValidTemplate (node_modules/eslint/lib/testers/rule-tester.js:396:28)
      at Context.RuleTester.it (node_modules/eslint/lib/testers/rule-tester.js:540:25)
```

**Comment by kevinzwhuang** [comment]
Thanks @ljharb! When do you think we'll get the next release with this fix in?
cc: @yannickcr
